packages/components/src/validated-form-controls/control-with-error.tsx

```diff
@@ -0,0 +1,161 @@
+import { Icon } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { error } from '@wordpress/icons';
+import { cloneElement, forwardRef, useEffect, useState } from 'react';
+import { withIgnoreIMEEvents } from '../utils/with-ignore-ime-events';
+
+import './style.scss';
+
+function appendRequiredIndicator(
+	label: string,
+	required: boolean | undefined,
+	markWhenOptional: boolean | undefined
+) {
+	if ( required && ! markWhenOptional ) {
+		return `${ label } (${ __( 'Required' ) })`;
+	}
+	if ( ! required && markWhenOptional ) {
+		return `${ label } (${ __( 'Optional' ) })`;
+	}
+	return label;
+}
+
+/**
+ * HTML elements that support the Constraint Validation API.
+ *
+ * Here, we exclude HTMLButtonElement because although it does technically support the API,
+ * normal buttons are actually exempted from any validation.
+ * @see https://developer.mozilla.org/en-US/docs/Learn_web_development/Extensions/Forms/Form_validation
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/HTMLButtonElement/willValidate
+ */
+type ValidityTarget =
+	| HTMLFieldSetElement
+	| HTMLInputElement
+	| HTMLSelectElement
+	| HTMLTextAreaElement;
+
+function UnforwardedControlWithError< C extends React.ReactElement >(
+	{
+		required,
+		markWhenOptional,
+		customValidator,
+		getValidityTarget,
+		render,
+	}: {
+		/**
+		 * Whether the control is required.
+		 */
+		required?: boolean;
+		/**
+		 * Label the control as "optional" when _not_ `required`, instead of the inverse.
+		 */
+		markWhenOptional?: boolean;
+		/**
+		 * A function that returns a custom validity message when applicable.
+		 *
+		 * This message will be applied to the element returned by `getValidityTarget`.
+		 * @see https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/setCustomValidity
+		 */
+		customValidator?: () => string | void;
+		/**
+		 * A function that returns the actual element on which the validity data should be applied.
+		 */
+		getValidityTarget: () => ValidityTarget | null | undefined;
+		render: C;
```

Comment:
nit: `render` feels like it should be a function that returns a React element, but it's currently a React element. Would it be better to rename it to `control` or `children`?

Reply:
Great point. I started out mimicking how Ariakit [adopts](https://ariakit.org/guide/composition) the `render` prop pattern, but seeing how my implementation ended up, I agree it kind of makes more sense as a `children` prop. Fixed.